src/user_crate/target.rs

```diff
@@ -0,0 +1,80 @@
+/*! PL/Rust adopts the tactic of always explicitly specifying which target to build for.
+    This prevents using the "fallback" logic of Cargo leaving builds in an unlabeled directory.
+    This is a precaution as PL/Rust is a cross-compiler.
+    so a normal build-and-test cycle may create artifacts for multiple targets.
+!*/
+
+use std::env;
+use std::ffi::OsString;
+use std::fmt::{self, Display, Formatter};
+
+pub(crate) mod host {
+    use std::env::consts::*;
+    cfg_if::cfg_if! { if #[cfg(target_env = "gnu")] {
+        pub(crate) const ENV: &str = "gnu";
+    } else if #[cfg(target_env = "musl")] {
+        pub(crate) const ENV: &str = "musl";
+    } else {
+        pub(crate) const ENV: &str = "";
+    }}
+    cfg_if::cfg_if! { if #[cfg(target_vendor = "apple")] {
+        pub(crate) const VENDOR: &str = "apple";
+    } else if #[cfg(target_os = "windows")] {
+        pub(crate) const VENDOR: &str = "pc";
+    } else {
+        pub(crate) const VENDOR: &str = "unknown";
+    }}
+
+    pub(crate) fn target_tuple() -> String {
+        let os = match OS {
+            "macos" => "darwin",
+            os => os
+        };
+        super::stringify_tuple([ARCH, VENDOR, os, ENV])
+    }
+}
+
+// Assemble a String from the components of a build tuple.
+fn stringify_tuple(tuple: [&str; 4]) -> String {
+    let mut s = String::from(tuple[0]);
+    for t in &tuple[1..] {
+        if t != &"" {
+            s.push('-');
+            s.push_str(t);
+        }
+    }
+    s
+}
+
+#[derive(Debug)]
+#[allow(dead_code)] // Such is the life of cfg code
```

Comment:
The rest of the code uses `thiserror` (see `src/error.rs`, would you be OK using that here as well?

Reply:
Yes, that would be fine!